cpp/src/arrow/compute/kernels/scalar_validity_test.cc

```diff
@@ -0,0 +1,151 @@
+  UnaryFunction func_;
+};
+
+TEST_F(TestValidityKernels, ArrayIsValid) {
+  func_ = arrow::compute::IsValid;
+
+  this->AssertUnary("[]", "[]");
+  this->AssertUnary("[null]", "[false]");
+  this->AssertUnary("[1]", "[true]");
+  this->AssertUnary("[null, 1, 0, null]", "[false, true, true, false]");
+}
+
+TEST_F(TestValidityKernels, ArrayIsValidBufferPassthruOptimization) {
+  Datum arg = ArrayFromJSON(boolean(), "[null, 1, 0, null]");
+  ASSERT_OK_AND_ASSIGN(auto validity, arrow::compute::IsValid(arg));
+  ASSERT_EQ(validity.array()->buffers[1], arg.array()->buffers[0]);
+}
+
+TEST_F(TestValidityKernels, ScalarIsValid) {
+  func_ = arrow::compute::IsValid;
+
+  AssertUnary(Datum(19.7), Datum(true));
+  AssertUnary(MakeNullScalar(float64()), Datum(false));
+}
+
+TEST_F(TestValidityKernels, ArrayIsNull) {
+  func_ = arrow::compute::IsNull;
+
+  this->AssertUnary("[]", "[]");
+  this->AssertUnary("[null]", "[true]");
+  this->AssertUnary("[1]", "[false]");
+  this->AssertUnary("[null, 1, 0, null]", "[true, false, false, true]");
+}
+
+TEST_F(TestValidityKernels, DISABLED_ScalarIsNull) {
+  func_ = arrow::compute::IsNull;
+
+  AssertUnary(Datum(19.7), Datum(false));
+  AssertUnary(MakeNullScalar(float64()), Datum(true));
+}
+
+class IsValidProperty : public ScalarFunctionPropertyMixin {
+ public:
+  std::shared_ptr<ScalarFunction> GetFunction() override {
+    return internal::checked_pointer_cast<ScalarFunction>(
+        *GetFunctionRegistry()->GetFunction("is_valid"));
+  }
+
+  Result<std::shared_ptr<Scalar>> Contract(const ScalarVector& args,
+                                           const FunctionOptions*) override {
+    return std::make_shared<BooleanScalar>(args[0]->is_valid);
+  }
+};
+
+TEST_P(IsValidProperty, TestIsValidProperty) { Validate(); }
+
+INSTANTIATE_TEST_SUITE_P(IsValidPropertyTest, IsValidProperty,
+                         ScalarFunctionPropertyTestParam::Values({
+                             {0, 0.0},
+                             {1, 0.0},
+                             {2, 0.0},
+                             {1024, 0.25},
+                         }));
+
+class IsNullProperty : public ScalarFunctionPropertyMixin {
+ public:
+  std::shared_ptr<ScalarFunction> GetFunction() override {
+    return internal::checked_pointer_cast<ScalarFunction>(
+        *GetFunctionRegistry()->GetFunction("is_null"));
+  }
+
+  Result<std::shared_ptr<Scalar>> Contract(const ScalarVector& args,
+                                           const FunctionOptions*) override {
+    return std::make_shared<BooleanScalar>(!args[0]->is_valid);
+  }
+};
+
+TEST_P(IsNullProperty, TestIsNullProperty) { Validate(); }
+
+INSTANTIATE_TEST_SUITE_P(IsNullPropertyTest, IsNullProperty,
+                         ScalarFunctionPropertyTestParam::Values({
+                             {0, 0.0},
+                             {1, 0.0},
+                             {2, 0.0},
+                             {1024, 0.25},
+                         }));
```

Comment:
Honestly this seems quite complex / opaque to me -- I am not going to be eager to write unit tests in this fashion. I would personally rather keep things as simple as possible.

Reply:
I agree that this looks over-engineered for what it seems to be doing (I may be misunderstanding) -- basically generating test data and checking expected values.

Reply:
I'll separate this into a separate PR, then. I think testing against random data has value, similar to using a fuzzer. Currently our random tests have a lot of boilerplate for generating the inputs and a lot of ad-hoc code for computing the expected values. IMHO it's worthwhile to have an interface for doing sanity checks across a wide swath of parameter space without needing to specify each of those manually (even if this is not enabled by default).

Reply:
You are free to open a separate PR but I'm against this particular implementation of this approach to testing so unless something materially changes about the way the developer specifies the tests I'm likely to vote against the PR

Reply:
https://issues.apache.org/jira/browse/ARROW-9135

Reply:
@wesm if you have any ideas which would make this interface simpler, could you comment in the JIRA?

Reply:
I'm not sure that the approach to testing is viable at all for this part of the project. The examples that were in this PR contained effectively reimplementations of the kernel logic in the property test specification. Let's look at Add for example

```
const auto& out_type = args[0]->type;

if (!args[0]->is_valid || !args[1]->is_valid) {
  return MakeNullScalar(out_type);
}

if (is_integer(out_type->id())) {
  ARROW_ASSIGN_OR_RAISE(auto lhs, Cast<UInt64Scalar>(args[0]));
  ARROW_ASSIGN_OR_RAISE(auto rhs, Cast<UInt64Scalar>(args[1]));
  return UInt64Scalar(lhs->value + rhs->value).CastTo(out_type);
}

if (is_floating(out_type->id())) {
  ARROW_ASSIGN_OR_RAISE(auto lhs, Cast<DoubleScalar>(args[0]));
  ARROW_ASSIGN_OR_RAISE(auto rhs, Cast<DoubleScalar>(args[1]));
  return DoubleScalar(lhs->value + rhs->value).CastTo(out_type);
}

return Status::NotImplemented("NYI");
```

I don't think it's a good idea to have a parallel collection of reimplementations of kernels.